File: Language/JavaScript/Behavioral/State/Example1/ConcreteStateA.js
// ConcreteStateA.js

const State = require('./State');

// ConcreteStateA handles the request
class ConcreteStateA extends State {
    handle(context) {
        console.log("ConcreteStateA handles request, switching to ConcreteStateB");
        context.setState(new (require('./ConcreteStateB'))()); // switch to B
    }
}

module.exports = ConcreteStateA;

File: TheRayCode/CPP/Behavioral/State/Example/ConcreteStateA.cpp
//
// Created by ray on 3/9/24.
//
#include "ConcreteStateA.h"
#include "ConcreteStateB.h"
#include "Context.h"
#include <iostream>

void ConcreteStateA::handle(Context* context) {
    std::cout << "State A handling context. Switching to State B." << std::endl;
    context->setState(new ConcreteStateB());
}

void ConcreteStateB::handle(Context* context) {
    std::cout << "State B handling context. Switching to State A." << std::endl;
    context->setState(new ConcreteStateA());
}